src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -46,29 +57,77 @@ uint16_t GenericOTARequestorDriver::GetMaxDownloadBlockSize()
     return 1024;
 }
 
-void GenericOTARequestorDriver::HandleError(UpdateFailureState state, CHIP_ERROR error)
+void StartDelayTimerHandler(System::Layer * systemLayer, void * appState)
+{
+    static_cast<GenericOTARequestorDriver *>(appState)->DriverSendQuery();
+}
+
+bool ProviderLocationsEqual(const ProviderLocation::Type & a, const ProviderLocation::Type & b)
 {
-    // TODO: Schedule the next QueryImage
+    if ((a.fabricIndex == b.fabricIndex) && (a.providerNodeID == b.providerNodeID) && (a.endpoint == b.endpoint))
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+void GenericOTARequestorDriver::HandleError(UpdateFailureState state, CHIP_ERROR error) {}
+
+void GenericOTARequestorDriver::HandleIdleState()
+{
+    // Default profivers timer runs if and only if the OTARequestor's update state is kIdle.
+    // Must (re)start the timer every time we enter the kIdle state
+    StartDefaultProviderTimer();
 }
 
 void GenericOTARequestorDriver::UpdateAvailable(const UpdateDescription & update, System::Clock::Seconds32 delay)
 {
+    // IMPLEMENTATION CHOICE:
+    // This implementation unconditionally downloads an available update
+
     VerifyOrDie(mRequestor != nullptr);
-    ScheduleDelayedAction(UpdateFailureState::kDownloading, delay,
-                          [](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); });
+    ScheduleDelayedAction(
+        delay, [](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); }, this);
 }
 
 void GenericOTARequestorDriver::UpdateNotFound(UpdateNotFoundReason reason, System::Clock::Seconds32 delay)
 {
     VerifyOrDie(mRequestor != nullptr);
 
+    ProviderLocation::Type providerLocation;
+
+    switch (reason)
+    {
+    case UpdateNotFoundReason::UpToDate:
+        return;
```

Comment:
Suggest replacing the early returns and empty break fallthroughs with a `bool willTryAnotherQuery` helper variable set to true/false in every single switch statement at the right place

Reply:
Done